src/python/pants/util/meta.py

```diff
@@ -111,7 +112,47 @@ def staticproperty(func: Union[staticmethod, Callable]) -> ClassPropertyDescript
   return ClassPropertyDescriptor(func, doc)
 
 
-def frozen_after_init(cls: Type[T]) -> Type[T]:
+class _ClassDecoratorWithSentinelAttribute(ABC):
+  """Base class to wrap a class decorator which sets a "sentinel attribute".
+
+  This functionality is exposed via the `@sentinel_attribute` decorator.
+  """
+
+  @abstractmethod
+  def __call__(self, cls: type) -> type: ...
```

Comment:
Technically should be `typing.Type` for the same reason it's `typing.List` instead of `list`

Reply:
What is that reason?

Reply:
To work with generics. You can't say `list[int]` or `type[Foo]`, but you can say `List[int]` and `Type[Foo]`.

However, PEP 585 is proposed to allow the former! https://www.python.org/dev/peps/pep-0585/ This is currently one of the most awkward parts of Python type hints.